Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -0,0 +1,479 @@
+{
+  std::vector<std::string> species_names = mpc->GetSpeciesNames();
+
+  for (unsigned i = 0, n = species_names.size(); i < n; ++i) {
+    auto& pc  = mpc->GetUniqueContainer(i);
+    if (pc->plot_species) {
+
+      amrex::Vector<std::string> real_names;
+      amrex::Vector<std::string> int_names;
+      amrex::Vector<int> int_flags;
+
+      real_names.push_back("weight");
+
+      real_names.push_back("momentum_x");
+      real_names.push_back("momentum_y");
+      real_names.push_back("momentum_z");
+
+      real_names.push_back("Ex");
+      real_names.push_back("Ey");
+      real_names.push_back("Ez");
+
+      real_names.push_back("Bx");
+      real_names.push_back("By");
+      real_names.push_back("Bz");
+
+#ifdef WARPX_DIM_RZ
+      real_names.push_back("theta");
+#endif
+      if(pc->do_field_ionization){
+         int_names.push_back("ionization_level");
+         // int_flags specifies, for each integer attribs, whether it is
+         // dumped to plotfiles. So far, ionization_level is the only
+         // integer attribs, and it is automatically dumped to plotfiles
+         // when ionization is on.
+         int_flags.resize(1, 1);
+      }
+
+      // Convert momentum to SI
+      pc->ConvertUnits(ConvertDirection::WarpX_to_SI);
+      // real_names contains a list of all real particle attributes.
+      // pc->plot_flags is 1 or 0, whether quantity is dumped or not.
+
+      {
+        //
+        SavePlotFile(pc,
+           species_names[i],
+           m_CurrentStep,
+           pc->plot_flags, // this is protected and accessible by MultiParticleContainer.
+           // so kept as is
+           int_flags,
+           real_names, int_names);
+      }
+
+      // Convert momentum back to WarpX units
+      pc->ConvertUnits(ConvertDirection::SI_to_WarpX);
+    }
+  }
+}
+
+
+
+void
+WarpXOpenPMDPlot::SavePlotFile (const std::unique_ptr<WarpXParticleContainer>& pc,
+                    const std::string& name,
+                    int iteration,
+                    const amrex::Vector<int>& write_real_comp,
+                    const amrex::Vector<int>& write_int_comp,
+                    const amrex::Vector<std::string>& real_comp_names,
+                    const amrex::Vector<std::string>&  int_comp_names) const
+{
+  if ( nullptr == m_Series)
+    return;
+
+  if (0 == m_CurrentStep ) // can only define once
+    m_Series->setParticlesPath("particles");
+
+  WarpXParticleCounter counter(pc);
+
+  openPMD::Iteration currIteration = m_Series->iterations[iteration];
+  openPMD::ParticleSpecies currSpecies = currIteration.particles[name];
+  //
+  // define positions & offsets
+  //
+  SetupPos(currSpecies, counter.GetTotalNumParticles());
+  SetupRealProperties(currSpecies, write_real_comp, real_comp_names, counter.GetTotalNumParticles());
+
+  for (auto currentLevel = 0; currentLevel <= pc->finestLevel(); currentLevel++)
+    {
+      long numParticles = counter.m_ParticleSizeAtRank[currentLevel];
```

Comment:
```suggestion
      long const numParticles = counter.m_ParticleSizeAtRank[currentLevel];
```
Or maybe even `unsigned long long const` as well?